### PR TITLE
Add dpnp.linalg.eigvalsh() implementation

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR adds a new `dpnp.linalg.eigvalsh()` function to calculate the eigenvalues of a complex Hermitian or real symmetric matrix using `dpnp.linalg.eigh()`
Also the PR adds missing tests for `dpnp.linalg.eigh` slightly changes its implementation and its docstrings

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: tests/test_linalg.py [hunk 1]
```diff
@@ -383,47 +383,32 @@ def test_eig_arange(type, size):
...
-    symm = symm_orig
-    dpnp_symm_orig = inp.array(symm)
-    dpnp_symm = dpnp_symm_orig
-
-    dpnp_val, dpnp_vec = inp.linalg.eigh(dpnp_symm)
-    np_val, np_vec = numpy.linalg.eigh(symm)
-
-    # DPNP sort val/vec by abs value
-    vvsort(dpnp_val, dpnp_vec, size, inp)
+    def test_eigenvalue_errors(self, func):
```
Comment (antonwolfy) [inline tests/test_linalg.py:394]: why do you remove a positive test scenario and replace it by negative one only?

File: tests/test_sycl_queue.py [hunk 2]
```diff
@@ -1087,43 +1087,71 @@ def test_eig(device):
...
-    dpnp_symm_orig = dpnp.array(numpy_data, device=device)
-    dpnp_data = dpnp_symm_orig
+    numpy.random.seed(81)
+    a = numpy.random.randn(*shape).astype(dtype)
+    if a.size > 0:
+        if a.ndim > 2:
+            for i in range(a.shape[0]):
+                a[i] = numpy.conj(a[i].T) @ a[i]
+        else:
+            a = numpy.conj(a.T) @ a
```
Comment (antonwolfy) [inline tests/test_sycl_queue.py:1128]: Looks as a common function which might be reused in other place to build Hermitian matrix. What do you think?

File: tests/test_sycl_queue.py [hunk 3]
```diff
@@ -1087,43 +1087,71 @@ def test_eig(device):
...
+    if func == "eigh":
+        dp_val, dp_vec = dpnp.linalg.eigh(dp_a)
+        np_val, np_vec = numpy.linalg.eigh(a)
 
-    expected_queue = dpnp_data.get_array().sycl_queue
-    dpnp_val_queue = dpnp_val.get_array().sycl_queue
-    dpnp_vec_queue = dpnp_vec.get_array().sycl_queue
+        assert_allclose(dp_vec, np_vec, rtol=1e-05, atol=1e-05)
+        assert dp_vec.shape == np_vec.shape
+
```
Comment (antonwolfy) [inline tests/test_sycl_queue.py:1140]: why did you remove assertion of dtypes?

File: tests/third_party/cupy/linalg_tests/test_eigenvalue.py [hunk 4]
```diff
@@ -28,7 +28,6 @@ class TestEigenvalue(unittest.TestCase):
         rtol=1e-3,
         atol=1e-4,
         type_check=has_support_aspect64(),
```
Comment (antonwolfy) [inline tests/third_party/cupy/linalg_tests/test_eigenvalue.py:30]: `unittest` can be also removed, since not used by cupy here

File: tests/third_party/cupy/linalg_tests/test_eigenvalue.py [hunk 5]
```diff
@@ -89,7 +90,9 @@ def test_eigh_batched(self, xp, dtype):
         return w
 
     @testing.for_complex_dtypes()
-    @testing.numpy_cupy_allclose(rtol=1e-3, atol=1e-4, contiguous_check=False)
+    @testing.numpy_cupy_allclose(
+        rtol=1e-3, atol=1e-4, type_check=has_support_aspect64()
+    )
```
Comment (antonwolfy) [inline tests/third_party/cupy/linalg_tests/test_eigenvalue.py:95]:
```suggestion
    @testing.numpy_cupy_allclose(rtol=1e-3, atol=1e-4)
```

File: tests/third_party/cupy/linalg_tests/test_eigenvalue.py [hunk 6]
```diff
@@ -139,19 +144,21 @@ def test_eigvalsh_batched(self, xp, dtype):
         # so they should be directly comparable
         return w
 
-    @pytest.mark.skip("No support of dpnp.eigvalsh()")
     @testing.for_complex_dtypes()
-    @testing.numpy_cupy_allclose(rtol=1e-3, atol=1e-4)
+    @testing.numpy_cupy_allclose(
+        rtol=1e-3, atol=1e-4, type_check=has_support_aspect64()
+    )
```
Comment (antonwolfy) [inline tests/third_party/cupy/linalg_tests/test_eigenvalue.py:150]:
```suggestion
    @testing.numpy_cupy_allclose(rtol=1e-3, atol=1e-4)
```

File: dpnp/linalg/dpnp_iface_linalg.py [hunk 7]
```diff
@@ -319,6 +322,63 @@ def eigvals(input):
...
+    array([0.17157288, 5.82842712])
+
+    """
+
+    dpnp.check_supported_arrays_type(a)
+    check_stacked_2d(a)
+    check_stacked_square(a)
+
+    if UPLO not in ("L", "U"):
+        raise ValueError("UPLO argument must be 'L' or 'U'")
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_iface_linalg.py:377]: numpy supports both upper and lower register of `L` and `F`, here and in `eigh`

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 8]
```diff
@@ -859,6 +853,21 @@ def dpnp_eigh(a, UPLO):
...
 
+def dpnp_eigvalsh(a, UPLO):
+    """
+    dpnp_eigvalsh(a, UPLO)
+
+    Return the eigenvalues of a complex Hermitian or real symmetric matrix.
+
+    """
+
+    if a.size == 0:
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply dpnp/linalg/dpnp_utils_linalg.py:864]: I decided to remove `dpnp_eigvalsh` and call `dpnp_eigh` instead

File: tests/test_sycl_queue.py (hunk 2 above)
Comment (vlad-perevezentsev) [reply tests/test_sycl_queue.py:1128]: I added a new `get_symm_herm_numpy_array` function to `helper.py`

File: tests/test_linalg.py (hunk 1 above)
Comment (vlad-perevezentsev) [reply tests/test_linalg.py:394]: Сause cupy tests cover all cases except errors and I did not want to duplicate the tests

File: dpnp/linalg/dpnp_iface_linalg.py [hunk 9]
```diff
@@ -277,20 +289,13 @@ def eigh(a, UPLO="L"):
     """
 
     dpnp.check_supported_arrays_type(a)
+    check_stacked_2d(a)
+    check_stacked_square(a)
 
+    UPLO = UPLO.upper()
     if UPLO not in ("L", "U"):
         raise ValueError("UPLO argument must be 'L' or 'U'")
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_iface_linalg.py:297]: the line is missing in coverage

File: dpnp/linalg/dpnp_iface_linalg.py [hunk 10]
```diff
@@ -322,6 +327,64 @@ def eigvals(input):
...
+
+    """
+
+    dpnp.check_supported_arrays_type(a)
+    check_stacked_2d(a)
+    check_stacked_square(a)
+
+    UPLO = UPLO.upper()
+    if UPLO not in ("L", "U"):
+        raise ValueError("UPLO argument must be 'L' or 'U'")
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_iface_linalg.py:383]: the line is missing in coverage

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 11]
```diff
@@ -733,72 +733,102 @@ def dpnp_det(a):
...
-        if op_count == 0:
-            return w, dpnp.empty_like(a, dtype=v_type)
-
-        eig_vecs = [None] * op_count
-        ht_copy_ev = [None] * op_count
-        ht_lapack_ev = [None] * op_count
-        for i in range(op_count):
+        batch_size = a.shape[0]
+        if batch_size == 0:
+            return (
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:789]: the line is missing in coverage

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 12]
```diff
@@ -733,72 +733,102 @@ def dpnp_det(a):
...
+                ht_list_ev[i], _ = getattr(li, lapack_func)(
+                    a_sycl_queue,
+                    jobz,
+                    uplo,
+                    a[i].get_array(),
+                    w[i].get_array(),
+                    depends=[],
+                )
+
+            ht_list_ev.wait()
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:813]:
```suggestion
            dpctl.SyclEvent.wait_for(ht_list_ev)
```

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 13]
```diff
@@ -733,72 +733,102 @@ def dpnp_det(a):
...
+                (w, dpnp.empty_like(a, dtype=v_type))
+                if eigen_mode == "V"
+                else w
+            )
+
+        eig_vecs = [None] * batch_size
+
+        # When `eigen_mode == "N"` (jobz == 0), OneMKL LAPACK does not overwrite the input array.
+        # If the input array 'a' is already F-contiguous and matches the target data type,
+        # we avoid unnecessary memory allocation and data copying.
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:799]:
```suggestion
        # we can avoid unnecessary memory allocation and data copying.
```

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 14]
```diff
@@ -733,72 +733,102 @@ def dpnp_det(a):
...
-        for i in range(op_count):
+        batch_size = a.shape[0]
+        if batch_size == 0:
+            return (
+                (w, dpnp.empty_like(a, dtype=v_type))
+                if eigen_mode == "V"
+                else w
+            )
+
+        eig_vecs = [None] * batch_size
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:795]: can be moved after the below if statement , since it isn't used inside there

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 15]
```diff
@@ -733,72 +733,102 @@ def dpnp_det(a):
...
+
+        # When `eigen_mode == "N"` (jobz == 0), OneMKL LAPACK does not overwrite the input array.
+        # If the input array 'a' is already F-contiguous and matches the target data type,
+        # we avoid unnecessary memory allocation and data copying.
+        if eigen_mode == "N" and a_order == "F" and a.dtype == v_type:
+            ht_list_ev = [None] * batch_size
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:801]: the block of code is missing in coverage

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 16]
```diff
@@ -807,29 +837,40 @@ def dpnp_eigh(a, UPLO):
...
+        ht_list_ev = []
+        copy_ev = None
 
-        # use DPCTL tensor function to fill the array of eigenvectors with content of input array
-        ht_copy_ev, copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
-            src=a_usm_arr, dst=v.get_array(), sycl_queue=a_sycl_queue
-        )
+        # When `eigen_mode == "N"` (jobz == 0), OneMKL LAPACK does not overwrite the input array.
+        # If the input array 'a' is already F-contiguous and matches the target data type,
+        # we avoid unnecessary memory allocation and data copying.
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:854]:
```suggestion
        # we can avoid unnecessary memory allocation and data copying.
```

File: dpnp/linalg/dpnp_utils_linalg.py [hunk 17]
```diff
@@ -807,29 +837,40 @@ def dpnp_eigh(a, UPLO):
...
+        if eigen_mode == "N" and a_order == "F" and a.dtype == v_type:
+            v = a
```
Comment (antonwolfy) [inline dpnp/linalg/dpnp_utils_linalg.py:856]: the line is missing in coverage

File: tests/helper.py [hunk 18]
```diff
@@ -166,6 +166,64 @@ def get_all_dtypes(
...
+        A random numpy array of the specified shape and dtype.
+        The array is Hermitian or symmetric if `hermitian` is True.
+
+    Note:
+    For arrays with more than 2 dimensions, the Hermitian or
+    symmetric property is ensured for each 2D sub-array.
+
+    """
+
+    if use_seed:
```
Comment (antonwolfy) [inline tests/helper.py:209]: what if to use `seed_value == None` instead of `use_seed` flag

**Review by antonwolfy — APPROVED** [review]
Thank you @vlad-perevezentsev
